Add Rubocop WIP

jaysonesmith opened the pull request:
![https://14415-presscdn-0-52-pagely.netdna-ssl.com/wp-content/uploads/2014/07/robocop-edit1.jpg](https://14415-presscdn-0-52-pagely.netdna-ssl.com/wp-content/uploads/2014/07/robocop-edit1.jpg)

Resolves nothing ..._yet_. Eventually, it'll resolve #120!

Everyone likes to have pretty code, right? Agreed! 

Anyway, Rubocop is 👍, and I'll be chipping away at the various violations that running `rubocop --auto-gen-config` created. Commits will come after each cop has been handled.

_Part man. Part machine. All cop. The future of Ruby enforcement._

jaysonesmith commented:
I've fixed all of the auto fixable issues and also at least one or two non-auto fixable cops. Now what's left to tackle are some of the more involved ones.

leesharma commented:
Wow, nice work!! 🎉 

I think it makes sense to merge the fixes in in multiple pieces (unless you're about done with the hard fixes). I'd like to wait a few days before doing that though; we've got a couple new team members coming on, and I'd like to hear their style preferences before we merge.

For now, I skimmed over the diffs and have a few thoughts/suggestions:
* We should exclude `db/schema.rb` from all checks. Actually, I'm not sure it's worth linting anything in this folder.
* Are you using the [Rails cops](https://rubocop.readthedocs.io/en/latest/cops_rails/)? Those'll probably be trickier fixes, but they'd be good to add in another PR.
* There are a few easy merge conflicts–people have been active with their PRs!

And also a few auto-correct bugs:
* 5b29189 Fix Layout/IndentationWidth - this unaligned a spec block, though it looks like 3f8b930 may have fixed it?
* a24607e Fix Layout/MultilineOperationIndentation - this unaligned some comments
* 498d574 Fix Style/NumericLiterals - shouldn't force special formatting for zipcodes/timestamps/etc.
* A few cops interact oddly with the specs, including 133c2c3 and 94e56e4

There are also a handful of default cops I disagree with, but I'll save that for the discussion later.

Again, awesome job! That string literal commit alone was daunting.

jaysonesmith commented:
@leesharma Thanks for the time and the review! 

I'm fine with waiting on merging. :) The changes are here and are ready to go whenever. I can begin work on the other cops in another branch so that we can break things up a bit more. (I realize there have been a ton of changes in this PR so far)

- I've reverted the changes to the db/ directory and added it to the exclude list in rubocop's yml. 
- I haven't had any experience with the Rails cops nor have I been doing anything to specifically call on them, but can do!
- I'll fix the merge conflicts and then stop dev on this branch, sound good?

I've gotta jet, for now, I'll take a look at your other points too. :) I also don't agree with all the cops, but for the most part, I'm fine. I don't like formatting numbers with underscores.

jaysonesmith commented:
- 5b29189 Fix Layout/IndentationWidth - this unaligned a spec block, though it looks like 3f8b930 may have fixed it?
  - Yeah, looks like it!
- a24607e Fix Layout/MultilineOperationIndentation - this unaligned some comments
  - Ah, I see what happened. It tabbed in the continuation of the URL that's being concatenated which pushed out the comments. That's an easy enough fix.
- 498d574 Fix Style/NumericLiterals - shouldn't force special formatting for zipcodes/timestamps/etc.
  - I totally agree on this one. I don't like seeing numbers with underscores in them, no matter what the numbers are! I've disabled that cop in `rubocop.yml`
- A few cops interact oddly with the specs, including 133c2c3 and 94e56e4
  - So, for `133c2c3`, that was for the BlockDelimeter cop. According to the [style guide](https://github.com/bbatsov/ruby-style-guide#single-line-blocks) `do ... end` shouldn't be used for single line blocks, but instead curly braces `{...}`. See the link for more info. Happy to revert if needed.
  - For `94e56e4`, Rubocop was trying to enforce [this syntax](https://github.com/bbatsov/ruby-style-guide#syntax) to help with what it considers ambiguous. It's strange that it doesn't throw up on this more frequently given this is all over Rspec's syntax. I guess perhaps it's having issues here because change takes a block and not_to takes that result? Can disable this one too if you'd like?

jaysonesmith commented:
Travis failed on the `rspec ./spec/features/managing_users_spec.rb:28` test that we'v ebeen working on. :(

leesharma commented:
Boo. At least it's a new failure message though... this might be another issue.

jaysonesmith commented:
Oh hey, that it is! Fun times. 🥇

jaysonesmith commented:
The failure is fixed, but there are a whole lotta merge conflicts now. I don't have the permissions to fix/see them, but I imagine there are style things that have been fixed in my work that are still present on master.

leesharma commented:
Yeah, I bet–this project's been really active this past week! You can solve conflicts by rebasing/merging with the `master` branch locally.

I'll open a PR on your branch with a merge resolving the conflicts. Feel free to use that or do it yourself later; whatever's easier for you.

jaysonesmith commented:
Thanks, @leesharma! We should be good to go now.

jaysonesmith commented:
Also @leesharma , I'll work on the Rails cops in another branch. :)

jaysonesmith commented:
Just have the linelength cop to do, but can do that after we get this first round in.

seanmarcia commented:
Awesome!!! :)
